### PR TITLE
Add support for `http.client_ip` tag for Rack-based frameworks

### PR DESCRIPTION
**Pull request by barweiss**
**What does this PR do?**

Add support for the `http.client_ip` tag for Rack-based API frameworks.

**Motivation**

The `http.client_ip` tag will help distinguish whether the request came from a public or a private address, allowing Datadog to distinguish external from internal services and API endpoints.

**How to test the change?**

The change is validated by RSpec behavior tests.

To test integration:

Set up and start a demo Rails server and enable auto-instrumentation.
Then use the following commands:

```bash
curl localhost:3000 # direct IP
curl -H 'X-Forwarded-For: 1.1.1.1' localhost:3000 # forwarded IP
curl -H 'X-Forwarded-For: 1.1.1.1' -H 'X-Forwarded-For: 1.1.1.1' localhost:3000 # two values for the header = no client ip
curl -H 'X-Forwarded-For: 1.1.1.1' -H 'X-Real-Ip: 1.2.3.4' localhost:3000 # two "ip" headers set = no client ip
```

Review the logs from the server. The emitted traces should reflect the expected results.

### CODE REVIEW

File: lib/datadog/core/configuration/settings.rb
Comment (lloeki) [inline lib/datadog/core/configuration/settings.rb:633]:
```suggestion

            # An optional name of a custom header to resolve the client IP from.
```

Linter should have caught this?

File: lib/datadog/tracing/client_ip.rb
Comment (lloeki) [inline lib/datadog/tracing/client_ip.rb:71]: I feel like this could go to the `Core` namespace.

File: lib/datadog/tracing/client_ip.rb
Comment (lloeki) [inline lib/datadog/tracing/client_ip.rb:26]: I feel like this could go to the `Core` namespace.

File: lib/datadog/tracing/client_ip.rb
Comment (lloeki) [inline lib/datadog/tracing/client_ip.rb:59]: I feel this nesting of `unless` is confusing.

Is this necessary since `extract_ip_from_full_address` seems to return the address when its internal criteria don't match?

File: lib/datadog/tracing/client_ip.rb
Comment (lloeki) [inline lib/datadog/tracing/client_ip.rb:86]: Looks like a job for `DEFAULT_IP_HEADERS_NAMES.each_with_object({}) do |name, result|`

File: lib/datadog/tracing/client_ip.rb
Comment (lloeki) [inline lib/datadog/tracing/client_ip.rb:99]:
```suggestion

        case ip_values_from_headers.size
```

Method return value of this method comes from the `case` return value, so I'd suggest to space this to conform with the Ruby style guide where implicit returns live on their own "line".

File: lib/datadog/tracing/client_ip.rb
Comment (lloeki) [reply lib/datadog/tracing/client_ip.rb:143]: If we decide to vendor third party code, we should find a way to slip the corresponding license, at least in `LICENSE-3rd_party.csv`.

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (lloeki) [inline lib/datadog/tracing/contrib/rack/middlewares.rb:230]: Similarly to above, looks like a job for `each_with_object`

File: lib/datadog/tracing/client_ip.rb
Comment (lloeki) [inline lib/datadog/tracing/client_ip.rb:53]: So this is only done when there's no IP reported? This feels off somehow (like, when would this happen), so I must be missing something.

**Review by lloeki — CHANGES_REQUESTED** [review]
A few changes needed but overall LGTM.

It seems like CircleCI ought to catch some linting issues but it did not run, or Rubocop is set to ignore them.

I'd rather have a second review from the `Tracing` team as well :)

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:143]: I avoided `IPAddr` initially because it has the netmask handling which I didn't want but I guess it's solvable with some verification and it spares vendoring from `ipaddress`.
I'll give `IPAddr` a shot

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:108]: Nice! Fixed

File: lib/datadog/core/configuration/settings.rb
Comment (barweiss) [reply lib/datadog/core/configuration/settings.rb:633]: Whoops.
Rubocop seems to pass but I shouldv'e considered the formatting in the other options.
Fixed

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:71]: Nice, fixed

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:26]: Fixed

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:59]: I completely refactored that part, let me know if it makes more sense now :)

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:99]: Good point, done

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (barweiss) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:230]: I was trying to modify this function as little as possible while still taking advantage of the new class I added :)

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:53]: Let me know if this is still confusing after the refacrtor

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [inline lib/datadog/tracing/client_ip.rb:39]: Noob question: should this have `!` in the name?

File: spec/datadog/tracing/client_ip_spec.rb
Comment (TonyCTHsu) [inline spec/datadog/tracing/client_ip_spec.rb:38]: I think this could be simplified, without custom matcher
```
describe ".validate_ip" do
  ips = [
    ...
  ]

  ips.each do |ip|
    context 'when given #{ip}' do
      let(:normalised) { described_class.strip_decorations(ip) }
      subject { described_class.validate_ip(normalised) }

      it { is_expected.to be true }
    end
  end
end
```

File: spec/datadog/tracing/client_ip_spec.rb
Comment (TonyCTHsu) [inline spec/datadog/tracing/client_ip_spec.rb:103]: `with(any_args)` is redundant and assertion `expect(span).to_not receive(:set_tag)` should be strong enough.

File: lib/datadog/tracing/client_ip.rb
Comment (barweiss) [reply lib/datadog/tracing/client_ip.rb:39]: Cool thanks!

As for using exceptions, I was indeed wondering whether this would have a performance impact. I'll revert it to a version that isn't using exceptions.

File: lib/datadog/core/configuration/settings.rb
Comment (barweiss) [reply lib/datadog/core/configuration/settings.rb:629]: I usually agree but since the environment variable is `DISABLED` I figured I'd keep it `disabled`.
Do you still think it's worth the setting name?

File: spec/datadog/tracing/client_ip_spec.rb
Comment (barweiss) [reply spec/datadog/tracing/client_ip_spec.rb:103]: Awesome, fixed

File: spec/datadog/tracing/client_ip_spec.rb
Comment (barweiss) [reply spec/datadog/tracing/client_ip_spec.rb:38]: Fixed

File: lib/datadog/core/configuration/settings.rb
Comment (TonyCTHsu) [reply lib/datadog/core/configuration/settings.rb:629]: Can we also rename the environment variable to `DD_TRACE_CLIENT_IP_HEADER_ENABLED` to be consistent with other environment variable?

**Comment by barweiss** [comment]
One of the remaining issues in this PR is choosing to use the builtin `IPAddr` class, or vendoring the `ipaddress` gem.

In order to check whether a string is a valid IP with `IPAddr`, we need to see that `IPAddr.new` does not raise an exception. However, `IPAddress.valid?` allows checking for invalid IP addresses without raising exceptions.

One of my concerns was whether raising an exceptions will have a significant performance impact, considering that in cases of misconfiguration, an invalid IP can be present in *every* traced request.

After a discussion with @TonyCTHsu, and following his advice, I ran some benchmarks that make it clear there's no significant performance hit:

```
Warming up --------------------------------------
              ipaddr    13.742k i/100ms
           ipaddress    14.983k i/100ms
Calculating -------------------------------------
              ipaddr    141.715k (± 1.5%) i/s -    714.584k in   5.043502s
           ipaddress    148.502k (± 1.6%) i/s -    749.150k in   5.046105s

Comparison:
           ipaddress:   148501.6 i/s
              ipaddr:   141714.9 i/s - 1.05x  (± 0.00) slower
```

**Review by TonyCTHsu — APPROVED** [review]

**Review by lloeki — APPROVED** [review]

**Comment by lloeki** [comment]
Approved, provided the integration CI failures are investigated if they still fail after a re-run.

File: lib/datadog/core/header_collection.rb
Comment (marcotc) [inline lib/datadog/core/header_collection.rb:31]: `Hash#transform_keys` is only available in Ruby 2.5 or later:
```ruby
puts({}.transform_keys(&:downcase))
#  == 2.4.10 ==
# -e:1:in `<main>': undefined method `transform_keys' for {}:Hash (NoMethodError)
# Did you mean?  transform_values
#
#  == 2.5.6 ==
# {}
```

Is this method being hit during our test runs? It should show up as a failure in 2.1-2.4 test cases, if so.

Comment (barweiss) [reply lib/datadog/core/header_collection.rb:31]: It's used in `spec/datadog/tracing/client_ip_spec`, odd. I'll try to see why it isn't called.

Comment (barweiss) [reply lib/datadog/core/header_collection.rb:31]: For some reason these tests don't fail in the pipeline:
https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/7179/workflows/73531d46-88fe-4910-af07-bb3d97451c23/jobs/265659/parallel-runs/1?filterBy=ALL
https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/7179/workflows/73531d46-88fe-4910-af07-bb3d97451c23/jobs/265659/parallel-runs/2?filterBy=ALL

**Review by marcotc — APPROVED** [review]

**Pull request merged**
